### PR TITLE
TST: Fix tolerance of tests for aarch64

### PR DESCRIPTION
**Pull request by siddhesh**
#### What does this implement/fix?

Some tests need slightly higher absolute tolerance in variation of
their results on aarch64 compared to the default.  The tolerances are
all still in the same order of magnitude.

### CODE REVIEW

**Comment by ilayn** [comment]
These would be better fixed to the machine precision of that architecture instead of fixed numbers. Then you don't need if branches for each architecture.

**Comment by siddhesh** [comment]
Sorry, could you please elaborate on how machine precision would help here?  Wouldn't it be a constant ~2.2e-16 on all IEEE754-conforming systems?

**Comment by ilayn** [comment]
The reason why you need different tolerance levels is not due to conformity but about architecture and the floating point operations. Hence we have to bound the precision of the results to the machine precision. 

These manually entered random numbers are simply some multiple of the machine precision and they should be adjusted such that even for nonconfoming architectures that might be tested later. 

For example, the first one is bounded by `0.005 * np.sqrt(np.spacing(1.))` and the others `500 * np.spacing(1.)` and `200*np.spacing(1.)`. The point is that you basically relax the coefficient of eps and not the absolute value. 

Otherwise we will get 20 if branches for every architecture which is very difficult to maintain later.

**Comment by pv** [comment]
I think the point is that the value of `np.spacing(1.)` is the same on all architectures, but the accuracy of computation varies.

**Comment by pv** [comment]
Rather than setting arch-specific tolerances, we have usually just taken a single tolerance that's big enough for all supported archs. So I'd suggest rounding the numbers and taking the largest one required. Here, the required increase seems to be below one order of magnitude which probably is OK.

**Comment by siddhesh** [comment]
Thanks for the explanation @ilayn , @pv . I understand the reason for differences in result precision, I just wasn't able to understand what that had to do with machine precision.  If I understand correctly, you're suggesting expressing error bounds as a function of machine precision, is that correct?

> Rather than setting arch-specific tolerances, we have usually just taken a single tolerance that's big enough for all supported archs. So I'd suggest rounding the numbers (e.g. to one significant digit) and taking the largest one required. Here, the required increase seems to be below one order of magnitude which probably is OK.

Got it, I'll just set the highest tolerance then since they're all of the same order.  Would it be more convenient to express them as a function of `np.spacing(1.)` as @ilayn suggested?

I have one more general question: how do I arrive at an error bounds function?  Are those documented somewhere or is it a standard definition (e.g. 1e-11 indicates order of `0.005 * np.sqrt(np.spacing(1.))`) that I just have to learn/know over time?  Thanks!

**Comment by ilayn** [comment]
That's usually decided when the algorithm is designed and benchmarked. For this particular case, filtfilt is basically doing two convolutions hence expected precision is half the float precision and therefore the sqrt. The extra coefficient is basically a matter of tolerance fitted manually, the smaller the better.

**Comment by rgommers** [comment]
This discussion seems to be getting a little complicated. Just bumping the number is what we normally do, no reason to do more here.

File: scipy/signal/tests/test_signaltools.py [hunk 1]
```diff
@@ -2080,9 +2081,12 @@ def filtfilt(self, zpk, x, axis=-1, padtype='odd', padlen=None,
             return sosfiltfilt(sos, x, axis, padtype, padlen)
 
     def test_basic(self):
+        atol = 1e-11
+        if platform.machine() == 'aarch64':
+            atol = 5.28e-11
         zpk = tf2zpk([1, 2, 3], [1, 2, 3])
         out = self.filtfilt(zpk, np.arange(12))
-        assert_allclose(out, arange(12), atol=1e-11)
+        assert_allclose(out, arange(12), atol=atol)
```
Comment (rgommers) [inline scipy/signal/tests/test_signaltools.py]: So for clarity: changing this to `atol=.28e-11` and removing the changes further up

**Comment by siddhesh** [comment]
Also cc @tylerjereddy. I noticed some of these tests I bumped tolerances for were marked xfail for ppc64le.  Perhaps these new tolerances may help ppc64le as well, or maybe it needs higher tolerance.

**Comment by rgommers** [comment]
Thanks for pointing that out @siddhesh. Would you mind pushing a new commit that removes those knownfails for `ppcle64`? If the tests pass or we can make them pass with another small tolerance bump, that would be nice. If it's more complicated, we can put the `xfail`s back and leave it for another time.

**Comment by siddhesh** [comment]
Added another commit as you requested @rgommers .  I couldn't spot the ppc64le test runner in the CI jobs, is it something that needs to be triggered manually?

File: scipy/special/tests/test_orthogonal.py [hunk 2]
```diff
@@ -729,11 +728,11 @@ def test_roots_genlaguerre():
 
     vgq(rootf(0.1), evalf(0.1), weightf(0.1), 0., np.inf, 5)
     vgq(rootf(0.1), evalf(0.1), weightf(0.1), 0., np.inf, 25, atol=1e-13)
-    vgq(rootf(0.1), evalf(0.1), weightf(0.1), 0., np.inf, 100, atol=1e-13)
+    vgq(rootf(0.1), evalf(0.1), weightf(0.1), 0., np.inf, 100, atol=1.57e-13)
```
Comment (rgommers) [inline scipy/special/tests/test_orthogonal.py]: This is failing on `ppcle64`, needs another tiny bump:
```
E       Mismatch: 0.02%
E       Max absolute difference: 1.59480718e-13
E       Max relative difference: 2.22044605e-16
```

**Comment by rgommers** [comment]
>  I couldn't spot the ppc64le test runner in the CI jobs, is it something that needs to be triggered manually?

It's the last entry in the build matrix on TravisCI

**Comment by siddhesh** [comment]
Bumped up tolerance for genlaguerre; I've squashed it into the first commit and kept the ppc64le markers commit separate.

**Comment by rgommers** [comment]
Awesome, all green, in it goes. Thanks @siddhesh!

**Comment by rgommers** [comment]
@tylerjereddy this may be useful and safe to backport for 1.4.0. Not strictly necessary, but probably nice to have for aarch64 packagers.

**Comment by mdhaber** [comment]
Thanks @siddhesh! If you'd like your name added to [THANKS.txt](https://github.com/scipy/scipy/blob/master/THANKS.txt), please make an entry [here](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.5.0*).

**Pull request merged**